### ISSUE TITLE
Validation of required fields during deserialization

### ISSUE THREAD

**Issue by PaulRReynolds** [FirelyTeam/firely-net-sdk#2273]
When deserializing JSON to FHIR resources, my understanding is that properties with a cardinality of 1..1 in the FHIR specification would throw a `DeserializationFailedException` if they were missing. For example, I'm working with `Endpoint` resources and expected the following to throw an exception since the `status` property is missing from the JSON. Am I misunderstanding the scope of validation during deserialization?

**Example**
```c#
try
{
    var json = "{\"resourceType\": \"Endpoint\", \"address\": \"https://my.client-system.com/fhir/notifications\"}";
    var options = new JsonSerializerOptions().ForFhir(typeof(Endpoint).Assembly);
    var endpoint = JsonSerializer.Deserialize<Endpoint>(json, options);
}
catch (DeserializationFailedException ex)
{
    // Should throw an exception
}
```

**Expected behavior**
I'd expect this to throw a `DeserializationFailedException` since the JSON is missing the `Endpoint.Status` property which has a cardinality of 1..1 

**Version used:**
 - FHIR Version: [e.g. R5]
 - Version: [e.g. 4.3.0]

**Comment by ewoutkramer** [comment]
I think you are right. The `[Cardinality]` attributes on the POCO should have caught this. 

Would you mind trying calling `Validate()` explicitly to see if that does catch it?

```
using Hl7.Fhir.Validation;

var endpoint = ...;
endpoint.Validate(recurse: true, narrativeValidation: NarrativeValidationKind.FhirXhtml);
```

**Comment by PaulRReynolds** [comment]
@ewoutkramer Yes, when I call Validate() it does catch the error, with the following message: 
"Element with minimum cardinality 1 cannot be null. At Endpoint.StatusElement."

**Comment by ewoutkramer** [comment]
Ah, well, then we've found the bug. The parser does a piece-by-piece validation (as soon as it deserializes a subtree, it validates that subtree), to avoid calling `Validate()` at then end (and thus having to traverse the tree again).  The logic for this piece-by-piece validation must be at fault here.

For now, you can keep on calling `Validate()`, and then disable the validation while parsing, to avoid doing it twice. You achieve this by setting the `Validator` property of the `FhirJsonPocoDeserializerSettings` to `null`.

**Comment by ewoutkramer** [comment]
Classic: the deserializer validates all elements that it encounters (going bottom up), but this means that elements that do not appear in the instance do not get validated. This is fine for *almost* all validations, but obviously not for cardinality validation - which needs to run on all elements with no values as well.

**Comment by PaulRReynolds** [comment]
@ewoutkramer Do you have an ETA on when this fix will be available in the Hl7.Fhir.R4 Nuget feed? And do you know if it'll be a hotfix to SDK v4, or only in v5?

Thanks again

**Comment by ewoutkramer** [comment]
It's available in 5.0, which we published last week (or before, time flies...).
